### ISSUE TITLE
fsdp2 support grad_dtype

### ISSUE THREAD

**Issue by liqing9399** [pytorch/pytorch#170648]
### Problem
The `grad_dtype` property of PyTorch tensors, which allows specifying gradient dtype for leaf tensors, is not preserved when applying `torch.distributed.fsdp.fully_shard`. This breaks precision control in distributed training.

cc @chauhang @penguinwu

**Comment by weifengpy** [comment]
@soulitzer can you share more context on grad_dtype? 

@liqing9399 do you have a repro? is the grad_dtype only set on model.parameters()? do you set it on any input tensors?

**Comment by liqing9399** [comment]
"When I use the default fp32 as grad_dtype, FSDP actually uses bfloat16. The reason is that grad_dtype is not preserved during FSDP parameter replacement."
https://github.com/pytorch/pytorch/blob/0fabc3ba44823f257e70ce397d989c8de5e362c1/torch/distributed/fsdp/_fully_shard/_fsdp_param.py#L589

``` python
import os
import torch
import torch.distributed as dist
import torch.nn as nn
import torch.multiprocessing as mp


def setup(rank, world_size):
    os.environ['MASTER_ADDR'] = 'localhost'
    os.environ['MASTER_PORT'] = '29500'
    dist.init_process_group("nccl", rank=rank, world_size=world_size)
    torch.cuda.set_device(rank)


def cleanup():
    dist.destroy_process_group()


def test_simple(rank, world_size):
    """最简单的 grad_dtype 测试"""
    setup(rank, world_size)

    try:
        # 1. 创建模型并设置 grad_dtype
        model = nn.Sequential(
            nn.Linear(10, 20),
            nn.ReLU(),
            nn.Linear(20, 10)
        ).cuda(rank)

        for name, param in model.named_parameters():
            current = getattr(param, 'grad_dtype', 'NOT_SET')
            print(f" original model: {name}: grad_dtype ({current})")

        # 2. 应用 FSDP2
        from torch.distributed.fsdp import fully_shard, MixedPrecisionPolicy
        param_dtype, reduce_dtype = (torch.bfloat16, torch.float32)
        mp_policy = MixedPrecisionPolicy(param_dtype=param_dtype, reduce_dtype=reduce_dtype)
        fully_shard(model[0], mp_policy=mp_policy)
        fully_shard(model[2], mp_policy=mp_policy)
        fully_shard(model, mp_policy=mp_policy)

        # 4. 测试梯度
        print(f"\n[Rank {rank}] Testing gradients...")
        model.train()
        x = torch.randn(4, 10, device=f'cuda:{rank}')
        loss = model(x).sum()
        loss.backward()

        for name, param in model.named_parameters():
            if param.grad is not None:
                expected = getattr(param, 'grad_dtype', None)
                actual = param.grad.dtype
                print(f"  {name}: grad.dtype ({actual}), grad_dtype: {expected}")

    finally:
        cleanup()


if __name__ == "__main__":
    world_size = 2

    print("=" * 60)
    print("Simple FSDP2 grad_dtype Test")
    print("=" * 60)

    mp.spawn(
        test_simple,
        args=(world_size,),
        nprocs=world_size,
        join=True
    )
```
